bittide/src/Bittide/SwitchDemoProcessingElement.hs

```diff
@@ -0,0 +1,128 @@
+  forall nodes dom.
+  ( HiddenClockResetEnable dom
+  , 1 <= nodes
+  ) =>
+  SNat nodes ->
+  -- | Local clock cycle counter
+  Signal dom (Unsigned 64) ->
+  -- | Incomming crossbar link
+  Signal dom (BitVector 64) ->
+  -- | Device DNA
+  Signal dom (Maybe (BitVector 96)) ->
+  -- | When to read from the crossbar link
+  Signal dom (Unsigned 64) ->
+  -- | How many cycles to read from crossbar link
+  Signal dom (Index (nodes * 3)) ->
+  -- | When to write to the crossbar link
+  Signal dom (Unsigned 64) ->
+  -- | How many cycles to read from crossbar link
+  Signal dom (Index (nodes * 3)) ->
+  ( -- \| Outgoing crossbar link
+    Signal dom (BitVector 64)
+  , -- \| Buffer output
+    Signal dom (Vec (nodes * 3) (BitVector 64))
+  )
+switchDemoPe SNat localCounter linkIn maybeDna readStart readCycles writeStart writeCycles =
+  (linkOut, buffer)
+ where
+  readStartLocked = regEn maxBound (peState .==. pure Idle) readStart
+  readCyclesLocked = regEn maxBound (peState .==. pure Idle) readCycles
+  writeStartLocked = regEn maxBound (peState .==. pure Idle) writeStart
+  writeCyclesLocked = regEn maxBound (peState .==. pure Idle) writeCycles
+
+  localData :: Signal dom (Vec 3 (BitVector 64))
+  localData = bundle (unbundle dnaVec :< (pack <$> localCounter))
+   where
+    dnaVec :: Signal dom (Vec 2 (BitVector 64))
+    dnaVec = bitCoerce . zeroExtend <$> dnaLocked
+    dnaLocked = fromMaybe 0xBAAB_BAAB_BAAB_BAAB_BAAB_BAAB <$> maybeDna
+
+  linkOut = stateToLinkOutput <$> peState <*> buffer <*> localData
+
+  stateToLinkOutput ::
+    SimplePeState nodes ->
+    Vec (nodes * 3) (BitVector 64) ->
+    Vec 3 (BitVector 64) ->
+    BitVector 64
+  stateToLinkOutput state buf locData =
+    case state of
+      Write i
+        | i < 3 -> locData !! i
+        | otherwise -> buf !! i
+      _ -> 0xABBA_ABBA_ABBA_ABBA
+
+  -- \| The buffer stores all the incomming bittide data. For the Bittide Switch demo,
+  -- each FPGA sends its DNA and local clock cycle counter, along with all received data.
+  -- The last FPGA will therefore receive all DNAs and local clock cycle counters.
+  -- Order of datapacket: DNA 32 msbs, DNA 64 lsbs, local counter.
+  -- Order of packets: Current FPGA first, followed by the one before. Comparing all
+  -- buffers would look like a moving window.
+  buffer :: Signal dom (Vec (nodes * 3) (BitVector 64))
+  buffer = bundle $ regEn initVal <$> enableVec <*> linkInVec
+   where
+    initVal = 0xABBA_ABBA_ABBA_ABBA
+    linkInVec = repeat linkIn
+
+    enableVec :: Vec (nodes * 3) (Signal dom Bool)
+    enableVec = unbundle $ go <$> peState
+     where
+      go :: SimplePeState nodes -> Vec (nodes * 3) Bool
+      go (Read x) = (== x) <$> indicesI
+      go _ = repeat False
+
+  peState =
+    register
+      Idle
+      ( update
+          <$> localCounter
+          <*> readStartLocked
+          <*> readCyclesLocked
+          <*> writeStartLocked
+          <*> writeCyclesLocked
+          <*> peState
+      )
```

Comment:
This causes an off-by-one error. I think you should do something like:

```suggestion
  prevPeState = register Idle peState
  
  peState = 
    update
      <$> localCounter
      <*> readStartLocked
      <*> readCyclesLocked
      <*> writeStartLocked
      <*> writeCyclesLocked
      <*> prevPeState
```

And work on `peState`. That way you actually read/write on cycle _n_ instead of _n+1_.